### ISSUE TITLE
[BUG] Out of memory

### ISSUE THREAD

**Issue by ms609** [revbayes/revbayes#552]
**Describe the bug**
When attempting to start an Mk analysis on a morphological dataset, I encounter an out-of-memory error (std:bad_alloc when run locally; on slurm, jobs fail with an oom message).

The same script runs okay for 73 other datasets (except a few where it segfaults, #551).  I'm requesting 250G of memory on the slurm machine.  The dataset is not particularly large (NTAX=151 NCHAR=75 + 42).

**To Reproduce**
Check out https://github.com/mk-nt/3707_by_ki and `source("mcmcmc.Rev")`

**Computer info**
Happens locally on Win 10 and on the Linux HPC, usign rb 1.2.4 and 1.2.5-devel

**Comment by bjoelle** [comment]
based on profiling it looks like a lot of the memory footprint comes from vector allocations within `PhyloCTMCSiteHomogeneous` in particular `correctionLikelihoods` in `PhyloCTMCSiteHomogeneousConditional<charType>::resizeLikelihoodVectors` l943
I don't know enough about what's going on in the PhyloCTMC to debug further

**Comment by davidcerny** [comment]
Yes, this is caused by the ascertainment bias correction for the presence of parsimony-informative characters only. Replacing `coding="informative"` by `coding="variable"` eliminates the problem. I've recently run into a similar issue, except that I was "only" getting unreasonably long runtimes instead of an unreasonably large memory footprint. It looks like the latter is due to characters with very high state numbers; when I restricted the analysis to characters with 7 or fewer states, the memory consumption reached a peak of about 40 GB when creating the model object, and stabilized at 21.5 GB once the MC^3 sampler started running.

I remember @hoehna telling me that the implementation of the MK-parsinf correction for multistate characters should be considered experimental. @wrightaprilm might have an insight into what RevBayes is trying to do there, conceptually.

**Comment by bjoelle** [comment]
see also #170 for a partial workaround

**Comment by ms609** [comment]
Thanks for the pointer to #170.  

As I'm already working with a copy of the `data` object, I think the key change to make is adding a line equivalent to `data_copy[i].removeExcludedCharacters()`?  
This still uses all of the 4GB of memory on my local machine, but I'll see whether this gets the script running on the HPC, once the job reaches the top of the queue.



I wonder whether it would be worth considering including this action as part of the default behaviour of  `data_copy[i].setNumStatesPartition()`, or whether there might be cases where a user wants to retain the excluded characters. 

I couldn't find any documentation on the `.setNumStatesPartition()` method, besides the example in one of the tutorials – unless I've missed something, it could be useful to describe how this function works and when it may be desirable to also remove the excluded characters.

**Comment by bjoelle** [comment]
yes, that's correct - I think it should help because a lot of the higher state partitions have very few characters
I tend to agree that this behaviour should probably be the default, or at the very least we need to update the tutorials to include it (I was myself not aware of that before I started looking into it)

**Comment by ms609** [comment]
Thanks; I've added a PR to update the tutorial for the moment, until a decision's reached as to whether to update the default behaviour – possibly there's a case where e.g. a user may wish to add some additional characters to a partition before calling `removeExcludedCharacters()`, which would justify the retention of two separate function calls?

**Comment by davidcerny** [comment]
@bjoelle No, this workaround unfortunately doesn't apply here, since the default behavior of `.setNumStatesPartition()` already _is_ to not only _exclude_ the characters with other than the specified number of states, but also to _remove_ them altogether. At the Rev level, a good way to check if a method is removing or just excluding characters is to run these two lines:

```
partition.nchar()                                                   # gives the total (i.e., included + excluded) num. of characters
partition.getIncludedCharacterIndices().size() # gives the number of included characters only
```

You'll note that for the partitions created in @ms609 's script (`matrixByState[k]`), `partition.nchar()` already yields the desired value. Accordingly, replacing

```
matrixByState[k].setNumStatesPartition(k)
```

by

```
matrixByState[k].setNumStatesPartition(k)
matrixByState[k].removeExcludedCharacters()
```

has no effect. Even when I inserted that extra line, I still had to kill the analysis once its memory usage exceeded ~90 GB...

@ms609 Because of the above, I would suggest closing your PR in the documentation repo; the addition is redundant.

**Comment by bredelings** [comment]
I'm not sure conditioning on parsimony-informative characters is going to work very well.  In order to compute the probability that a character is parsimony informative, you basically have to create a fake data set that contains all parsimony-informative patterns -- or all uninformative patterns, whichever is smaller.  Then you can sum up the column probabilities.  Is there a clever way of doing this more cheaply?  I'm not sure...

If I'm thinking about this correctly, this can be a truly huge number of patterns!  The number of patterns in the fake data set for variable characters is much lower.

You can share the probability of something being parsimony informative across all columns with no gaps.  But if a column has gaps, then you have to create another fake data set for it.  

In theory, if you know that the rate matrix has symmetries, like a Jukes-Cantor matrix, you could use a lot fewer patterns.

Caveat: I still haven't looked at the code in revbayes for this yet.

**Comment by wrightaprilm** [comment]
I think that's more or less what the ascertainment bias correction is doing. It's here: https://github.com/revbayes/revbayes/blob/development/src/core/distributions/phylogenetics/substitution/PhyloCTMCSiteHomogeneousConditional.h

It's been a moment since I've thought about it, and I'm picking back through, since the hurricane is upon me and there's nothing else to do.

**Comment by davidcerny** [comment]
> I'm not sure conditioning on parsimony-informative characters is going to work very well.

I successfully used it [here](https://onlinelibrary.wiley.com/doi/10.1111/cla.12583), but in hindsight, it makes sense it worked okay for that particular dataset: (1) it had few taxa, (2) most characters were binary rather than multistate, and (3) even the multistate characters all had at most 4 states. These are pretty much the necessary conditions for the correction to be practical, as detailed in the appendix to [this paper](https://peerj.com/articles/4553/).

**Comment by davidcerny** [comment]
Maybe we could use the ratio of observed over simulated patterns to output a warning about the expected slowdown and memory footprint increase of an analysis when `coding="informative"`...

**Comment by ms609** [comment]
A warning message is a good idea; when I first came across the parsimony-informative correction I didn't put any thought into how it worked and had blithely assumed it was equivalent as the variable correction – I wondered why I hadn't seen it used more widely – it took me a while to recognize the impact it was having on my run times.  I think the philosophical justification for using the correction makes it worth exploring, even if in practice it seldom makes much difference (itself a somewhat surprising observation)

This said, the situation is perhaps not so bad as to make the current implementation unworkable – I've managed to get good results for 60+ of the sample of 75 real-world matrices I've been analysing in reasonable times, so I don't think we need to be too downbeat.

The problem of its calculation has the air of a problem for which an elegant solution, or at least simplification, exists – though my idle explorations have thus far yielded little more than a headache...

**Comment by ms609** [comment]
To follow up on this, I've been using `dnPhyloCTMC(coding = "informative")` with discrete characters that have multiple (>2) states.  RevBayes hasn't given any indication of a problem, and the results are not prominently dissimilar to those with `coding = "variable"`.

What I take from the discussion above is that whilst the current implementation may have room for further optimization, there is no reason to doubt its correctness.

The `coding` argument is not yet documented at [dnPhyloCTMC](https://revbayes.github.io/documentation/dnPhyloCTMC.html). A [tutorial](https://revbayes.github.io/tutorials/morph_tree/) that describes its use doesn't articulate any limitations (e.g. that it is only implemented for `fnJC(2)` – indeed, it advertises the RevBayes implementation as being equivalent to but faster than previous approaches.
Is anyone aware of a reason that `dnPhyloCTMC(coding = "informative")` would not behave as expected for characters with >2 states?
